panic on unknown log format

1. What did you do?

Use `--log-format=xml`. 

2. What did you expect to see?

Either an XML formatted log or an error saying only text and json are supported

3. What did you see instead?

```
[dvaneeden@dve-carbon test_db]$ tiup br backup full -s local:///tmp/tidb_backup_003 --log-format=xml
Starting component `br`: /home/dvaneeden/.tiup/components/br/v5.0.1/br backup full -s local:///tmp/tidb_backup_003 --log-format=xml
Detail BR log in /tmp/br.log.2021-05-04T16.48.04+0200 
panic: unsupport log format: xml

goroutine 1 [running]:
github.com/pingcap/log.NewTextEncoder(0xc0008383c0, 0xc0008441e0, 0x4)
	github.com/pingcap/log@v0.0.0-20201112100606-8f1e84a3abc8/zap_text_encoder.go:168 +0x3fb
github.com/pingcap/log.newZapTextEncoder(...)
	github.com/pingcap/log@v0.0.0-20201112100606-8f1e84a3abc8/config.go:78
github.com/pingcap/log.InitLoggerWithWriteSyncer(0xc0008383c0, 0x3a4b860, 0xc00084e0b0, 0x0, 0x0, 0x0, 0xc0006bba30, 0x1, 0x1, 0x37)
	github.com/pingcap/log@v0.0.0-20201112100606-8f1e84a3abc8/log.go:63 +0xc0
github.com/pingcap/log.InitLogger(0xc0008383c0, 0x0, 0x0, 0x0, 0x1, 0xc0007b3380, 0x37, 0x0)
	github.com/pingcap/log@v0.0.0-20201112100606-8f1e84a3abc8/log.go:53 +0xa7
main.Init.func1()
	github.com/pingcap/br@/cmd/br/cmd.go:115 +0x260
sync.(*Once).doSlow(0x57d9a78, 0xc0006bbb80)
	sync/once.go:66 +0xe3
sync.(*Once).Do(...)
	sync/once.go:57
main.Init(0xc000840dc0, 0x0, 0x0)
	github.com/pingcap/br@/cmd/br/cmd.go:89 +0xbc
main.NewBackupCommand.func1(0xc000840dc0, 0xc00084a2d0, 0x0, 0x3, 0x0, 0x0)
	github.com/pingcap/br@/cmd/br/backup.go:74 +0x2f
github.com/spf13/cobra.(*Command).execute(0xc000840dc0, 0xc0000f2150, 0x3, 0x3, 0xc000840dc0, 0xc0000f2150)
	github.com/spf13/cobra@v1.0.0/command.go:821 +0x56b
github.com/spf13/cobra.(*Command).ExecuteC(0xc000205340, 0xc0006bbe38, 0x3, 0x3)
	github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
	github.com/pingcap/br@/cmd/br/main.go:56 +0x35c
Error: run `/home/dvaneeden/.tiup/components/br/v5.0.1/br` (wd:/home/dvaneeden/.tiup/data/SWSD09N) failed: exit status 2
```


4. What version of BR and TiDB/TiKV/PD are you using?
```
[dvaneeden@dve-carbon test_db]$ tiup br --version
Starting component `br`: /home/dvaneeden/.tiup/components/br/v5.0.1/br --version
Release Version: v5.0.1
Git Commit Hash: b3bd5495369244cfb1d3f2c49763a7fdb55d4ecd
Git Branch: heads/refs/tags/v5.0.1
Go Version: go1.13
UTC Build Time: 2021-04-23 06:02:57
Race Enabled: false
```

Improvements:

- Add documentation for `--log-format`
- List possible choices in `--help` output
- Return a more helpful error message instead of a panic
- Maybe use "text" as fallback?

--- a/log.go
+++ b/log.go
@@ -23,7 +23,7 @@ import (
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 	"go.uber.org/zap/zaptest"
-	lumberjack "gopkg.in/natefinch/lumberjack.v2"
+	"gopkg.in/natefinch/lumberjack.v2"
 )
 
 var globalMu sync.Mutex
@@ -85,7 +85,10 @@ func InitLoggerWithWriteSyncer(cfg *Config, output, errOutput zapcore.WriteSynce
 	if err != nil {
 		return nil, nil, err
 	}
-	encoder := NewTextEncoder(cfg)
+	encoder, err := NewTextEncoder(cfg)
+	if err != nil {
+		return nil, nil, err
+	}
 	registerOnce.Do(func() {
 		err = zap.RegisterEncoder(ZapEncodingName, func(zapcore.EncoderConfig) (zapcore.Encoder, error) {
 			return encoder, nil

--- a/log_test.go
+++ b/log_test.go
@@ -20,7 +20,7 @@ import (
 	"net/url"
 	"testing"
 
-	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
@@ -87,12 +87,13 @@ func TestZapTextEncoder(t *testing.T) {
 
 	var buffer bytes.Buffer
 	writer := bufio.NewWriter(&buffer)
-	encoder := NewTextEncoder(conf)
+	encoder, err := NewTextEncoder(conf)
+	require.NoError(t, err)
 	logger := zap.New(zapcore.NewCore(encoder, zapcore.AddSync(writer), zapcore.InfoLevel)).Sugar()
 
 	logger.Info("this is a message from zap")
 	_ = writer.Flush()
-	assert.Equal(t, `[INFO] ["this is a message from zap"]`+"\n", buffer.String())
+	require.Equal(t, `[INFO] ["this is a message from zap"]`+"\n", buffer.String())
 }
 
 func TestRegisteredTextEncoder(t *testing.T) {
@@ -105,10 +106,10 @@ func TestRegisteredTextEncoder(t *testing.T) {
 	lgc.OutputPaths = []string{"memory://"}
 
 	lg, err := lgc.Build()
-	assert.Nil(t, err)
+	require.Nil(t, err)
 
 	lg.Info("this is a message from zap")
-	assert.Contains(t, sink.String(), `["this is a message from zap"]`)
+	require.Contains(t, sink.String(), `["this is a message from zap"]`)
 }
 
 // testingSink implements zap.Sink by writing all messages to a buffer.

--- a/zap_log_test.go
+++ b/zap_log_test.go
@@ -25,7 +25,7 @@ import (
 	"unsafe"
 
 	"github.com/pingcap/errors"
-	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
@@ -118,28 +118,30 @@ func TestLog(t *testing.T) {
 			`<car><mirror>XML</mirror></car>]"] [duration=10s]`,
 	)
 
-	assert.PanicsWithValue(t, "unknown", func() { sugar.Panic("unknown") })
+	require.PanicsWithValue(t, "unknown", func() { sugar.Panic("unknown") })
 }
 
 func TestTimeEncoder(t *testing.T) {
 	sec := int64(1547192741)
 	nsec := int64(165279177)
 	as, err := time.LoadLocation("Asia/Shanghai")
-	assert.Nil(t, err)
+	require.NoError(t, err)
 
 	tt := time.Unix(sec, nsec).In(as)
 	conf := &Config{Level: "debug", File: FileLogConfig{}, DisableTimestamp: true}
-	enc := NewTextEncoder(conf).(*textEncoder)
-	DefaultTimeEncoder(tt, enc)
-	assert.Equal(t, "2019/01/11 15:45:41.165 +08:00", enc.buf.String())
+	enc, err := NewTextEncoder(conf)
+	require.NoError(t, err)
+	DefaultTimeEncoder(tt, enc.(*textEncoder))
+	buf := enc.(*textEncoder).buf
+	require.Equal(t, "2019/01/11 15:45:41.165 +08:00", buf.String())
 
-	enc.buf.Reset()
+	buf.Reset()
 	utc, err := time.LoadLocation("UTC")
-	assert.Nil(t, err)
+	require.NoError(t, err)
 
 	utcTime := tt.In(utc)
-	DefaultTimeEncoder(utcTime, enc)
-	assert.Equal(t, "2019/01/11 07:45:41.165 +00:00", enc.buf.String())
+	DefaultTimeEncoder(utcTime, enc.(*textEncoder))
+	require.Equal(t, "2019/01/11 07:45:41.165 +00:00", buf.String())
 }
 
 // See [logger-header]https://github.com/tikv/rfcs/blob/master/text/2018-12-19-unified-log-format.md#log-header-section.
@@ -156,13 +158,15 @@ func TestZapCaller(t *testing.T) {
 		"ztest_coordinator1.go:20",
 		"<unknown>",
 	}
-	conf := &Config{Level: "deug", File: FileLogConfig{}, DisableTimestamp: true}
-	enc := NewTextEncoder(conf).(*textEncoder)
+	conf := &Config{Level: "debug", File: FileLogConfig{}, DisableTimestamp: true}
+	enc, err := NewTextEncoder(conf)
+	require.NoError(t, err)
 
 	for i, d := range data {
-		ShortCallerEncoder(d, enc)
-		assert.Equal(t, expect[i], enc.buf.String())
-		enc.buf.Reset()
+		ShortCallerEncoder(d, enc.(*textEncoder))
+		buf := enc.(*textEncoder).buf
+		require.Equal(t, expect[i], buf.String())
+		buf.Reset()
 	}
 }
 
@@ -176,7 +180,7 @@ func TestRotateLog(t *testing.T) {
 		},
 	}
 	logger, _, err := InitLogger(conf)
-	assert.Nil(t, err)
+	require.NoError(t, err)
 
 	var data []byte
 	for i := 1; i <= 1*1024*1024; i++ {
@@ -188,7 +192,7 @@ func TestRotateLog(t *testing.T) {
 		data = data[:0]
 	}
 	files, _ := os.ReadDir(tempDir)
-	assert.Len(t, files, 2)
+	require.Len(t, files, 2)
 	_ = os.RemoveAll(tempDir)
 }
 
@@ -227,8 +231,8 @@ func TestLogJSON(t *testing.T) {
 		"backoff", time.Second,
 	)
 	logger.With(zap.String("connID", "1"), zap.String("traceID", "dse1121")).Info("new connection")
-	ts.assertMessages("{\"level\":\"INFO\",\"caller\":\"zap_log_test.go:224\",\"message\":\"failed to fetch URL\",\"url\":\"http://example.com\",\"attempt\":3,\"backoff\":\"1s\"}",
-		"{\"level\":\"INFO\",\"caller\":\"zap_log_test.go:229\",\"message\":\"new connection\",\"connID\":\"1\",\"traceID\":\"dse1121\"}")
+	ts.assertMessages("{\"level\":\"INFO\",\"caller\":\"zap_log_test.go:228\",\"message\":\"failed to fetch URL\",\"url\":\"http://example.com\",\"attempt\":3,\"backoff\":\"1s\"}",
+		"{\"level\":\"INFO\",\"caller\":\"zap_log_test.go:233\",\"message\":\"new connection\",\"connID\":\"1\",\"traceID\":\"dse1121\"}")
 }
 
 // testLogSpy is a testing.TB that captures logged messages.
@@ -270,23 +274,23 @@ func (t *testLogSpy) Logf(format string, args ...interface{}) {
 }
 
 func (t *testLogSpy) assertMessages(msgs ...string) {
-	assert.Equal(t.TB, msgs, t.Messages)
+	require.Equal(t.TB, msgs, t.Messages)
 }
 
 func (t *testLogSpy) assertMessagesContains(msg string) {
 	for _, actualMsg := range t.Messages {
-		assert.Contains(t.TB, actualMsg, msg)
+		require.Contains(t.TB, actualMsg, msg)
 	}
 }
 
 func (t *testLogSpy) assertLastMessageContains(msg string) {
-	assert.NotEmpty(t.TB, t.Messages)
+	require.NotEmpty(t.TB, t.Messages)
 	lastMsg := t.Messages[len(t.Messages)-1]
-	assert.Contains(t.TB, lastMsg, msg)
+	require.Contains(t.TB, lastMsg, msg)
 }
 
 func (t *testLogSpy) assertMessagesNotContains(msg string) {
 	for _, actualMsg := range t.Messages {
-		assert.NotContains(t.TB, actualMsg, msg)
+		require.NotContains(t.TB, actualMsg, msg)
 	}
 }

--- a/zap_text_encoder.go
+++ b/zap_text_encoder.go
@@ -137,7 +137,7 @@ type textEncoder struct {
 
 // NewTextEncoder creates a fast, low-allocation Text encoder. The encoder
 // appropriately escapes all field keys and values.
-func NewTextEncoder(cfg *Config) zapcore.Encoder {
+func NewTextEncoder(cfg *Config) (zapcore.Encoder, error) {
 	cc := zapcore.EncoderConfig{
 		// Keys can be anything except the empty string.
 		TimeKey:        "time",
@@ -162,11 +162,11 @@ func NewTextEncoder(cfg *Config) zapcore.Encoder {
 			buf:                 _pool.Get(),
 			spaced:              false,
 			disableErrorVerbose: cfg.DisableErrorVerbose,
-		}
+		}, nil
 	case "json":
-		return zapcore.NewJSONEncoder(cc)
+		return zapcore.NewJSONEncoder(cc), nil
 	default:
-		panic(fmt.Sprintf("unsupport log format: %s", cfg.Format))
+		return nil, fmt.Errorf("unsupport log format: %s", cfg.Format)
 	}
 }
 

